x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/parser/AbstractBuilder.java

```diff
@@ -137,50 +133,76 @@ public static String unquoteString(Source source) {
         checkForSingleQuotedString(source, text, 0);
 
         text = text.substring(1, text.length() - 1);
-        StringBuffer resultString = new StringBuffer();
-        Matcher regexMatcher = slashPattern.matcher(text);
-
-        while (regexMatcher.find()) {
-            String group = regexMatcher.group();
-            String replacement;
-
-            switch (group) {
-                case "\\t":
-                    replacement = "\t";
-                    break;
-                case "\\b":
-                    replacement = "\b";
-                    break;
-                case "\\f":
-                    replacement = "\f";
-                    break;
-                case "\\n":
-                    replacement = "\n";
-                    break;
-                case "\\r":
-                    replacement = "\r";
-                    break;
-                case "\\\"":
-                    replacement = "\"";
-                    break;
-                case "\\'":
-                    replacement = "'";
-                    break;
-                case "\\\\":
-                    // will be interpreted as regex, so we have to escape it
-                    replacement = "\\\\";
-                    break;
-                default:
-                    // unknown escape sequence, pass through as-is
-                    replacement = group;
+        StringBuilder sb = new StringBuilder();
+
+        for (int i = 0; i < text.length();) {
+            if (text.charAt(i) == '\\') {
+                switch (text.charAt(++i)) {
+                    case 't':
+                        sb.append('\t');
+                        break;
+                    case 'b':
+                        sb.append('\b');
+                        break;
+                    case 'f':
+                        sb.append('\f');
+                        break;
+                    case 'n':
+                        sb.append('\n');
+                        break;
+                    case 'r':
+                        sb.append('\r');
+                        break;
+                    case '"':
+                        sb.append('\"');
+                        break;
+                    case '\'':
+                        sb.append('\'');
+                        break;
+                    case 'u':
+                        i = handleUnicodeChars(source, sb, text, ++i);
+                        break;
+                    case '\\':
+                        sb.append('\\');
+                        // will be interpreted as regex, so we have to escape it
+                        break;
+                    default:
```

Comment:
not sure if this can happen (as it won't be a `STRING_ESCAPE` char?), but prolly safer this way.

Reply:
If instead of passthrough I through an exception:
```
throw new ParsingException("lala");
```
and this test that I added with this PR won't work:
```
assertEquals("hello\\world", unquoteString(source("\"hello\\world\"")));
```